server/text_generation_server/utils/weights.py

```diff
@@ -1,13 +1,66 @@
-import os
+from abc import ABC, abstractmethod
 from pathlib import Path
 from typing import Dict, List, Optional, Union
-from safetensors import safe_open, SafetensorError
+from safetensors import safe_open
 import torch
-from loguru import logger
-from huggingface_hub import hf_hub_download
-import json
-from text_generation_server.layers.gptq import GPTQParams
-from text_generation_server.utils.log import log_once
+
+
+class WeightsLoader(ABC):
+    """
+    Instances of this type implement higher-level weight loading.
+
+    At a low-level, every weight is stored in the Safetensors format.
+    The interpretation of weights may be different however, for instance
+    could be packed, quantized weights. Loaders are responsible for
+    interpreting the raw tensors, sharding tensors in a manner compatible
+    with the format, etc.
+    """
+
+    @abstractmethod
+    def get_weights_col_packed(
+        self,
+        weights: "Weights",
+        prefix: str,
+        block_sizes: Union[int, List[int]],
+    ): ...
+
+    @abstractmethod
+    def get_weights_col(self, weights: "Weights", prefix: str): ...
+
+    @abstractmethod
+    def get_multi_weights_col(
+        self, weights: "Weights", prefixes: List[str], dim: int
+    ): ...
+
+    @abstractmethod
+    def get_multi_weights_row(self, weights: "Weights", prefix: str): ...
+
+
+class _DefaultWeightsLoader(WeightsLoader):
+    """
+    This loader uses tensors as-is with the exception of applying sharding
+    and/or concatenation.
+    """
+
+    def get_weights_col_packed(
+        self,
+        weights: "Weights",
+        prefix: str,
+        block_sizes: Union[int, List[int]],
+    ):
+        return weights.get_packed_sharded(
+            f"{prefix}.weight", dim=0, block_sizes=block_sizes
+        )
+
+    def get_weights_col(self, weights: "Weights", prefix: str):
+        return weights.get_multi_weights_col([prefix], 0)
```

Comment:
Isn't that always equivalent to `self.get_multi_weights_col([prefix], 0)` (meaning this could be directly into the core ABC interface ? leaving 1 thing less to implement ?.

Reply:
Except in Exl2, but we could indeed move the standard implementation to the ABC and only override it in the Exl2 loader.

Reply:
Fixed.